packages/flutter/lib/src/material/date_picker.dart

```diff
@@ -277,44 +416,111 @@ class _DatePickerDialog extends StatefulWidget {
   /// The initial display of the calendar picker.
   final DatePickerMode initialCalendarMode;
 
+  /// The error text displayed if the entered date is not in the correct format.
   final String? errorFormatText;
 
+  /// The error text displayed if the date is not valid.
+  ///
+  /// A date is not valid if it is earlier than [firstDate], later than
+  /// [lastDate], or doesn't pass the [selectableDayPredicate].
   final String? errorInvalidText;
 
+  /// The hint text displayed in the [TextField].
+  ///
+  /// If this is null, it will default to the date format string. For example,
+  /// 'mm/dd/yyyy' for en_US.
   final String? fieldHintText;
 
+  /// The label text displayed in the [TextField].
+  ///
+  /// If this is null, it will default to the words representing the date format
+  /// string. For example, 'Month, Day, Year' for en_US.
   final String? fieldLabelText;
 
   @override
   _DatePickerDialogState createState() => _DatePickerDialogState();
 }
 
-class _DatePickerDialogState extends State<_DatePickerDialog> {
+// A restorable data structure to save and restore dates. It only accounts
+// for the year, month and day.
+class _RestorableDate extends RestorableProperty<DateTime?> {
+  DateTime? _date;
 
-  late DatePickerEntryMode _entryMode;
-  late DateTime _selectedDate;
-  late bool _autoValidate;
-  final GlobalKey _calendarPickerKey = GlobalKey();
-  final GlobalKey<FormState> _formKey = GlobalKey<FormState>();
+  void _setDate(DateTime dateTime) {
+    _date = dateTime;
+    notifyListeners();
+  }
+
+  DateTime? _getDate() => _date;
 
   @override
-  void initState() {
-    super.initState();
-    _entryMode = widget.initialEntryMode;
-    _selectedDate = widget.initialDate;
-    _autoValidate = false;
+  DateTime? createDefaultValue() => null;
+
+  @override
+  DateTime? fromPrimitives(Object? data) {
+    if (data != null) {
+      final Map<dynamic, dynamic> tempData = data as Map<dynamic, dynamic>;
+      _date = DateTime(
+        tempData['year'] as int,
+        tempData['month'] as int,
+        tempData['day'] as int,
+      );
+    }
+
+    return _date;
+  }
+
+  @override
+  void initWithValue(DateTime? value) {
+    _date = value;
+  }
+
+  @override
+  Object toPrimitives() {
+    return <String, int?>{
+      'year': _date?.year,
+      'month': _date?.month,
+      'day': _date?.day,
+    };
+  }
```

Comment:
```suggestion
Object toPrimitives() => _date?.millisecondsSinceEpoch;
```

and then `fromPrimitives` would just be:

```dart
DateTime? fromPrimitives(Object? data) => data == null ? null : DateTime.fromMillisecondsSinceEpoch(data);
```